keps/sig-auth/4412-projected-service-account-tokens-for-kubelet-image-credential-providers/README.md

```diff
@@ -0,0 +1,1123 @@
+long lived, persisted in the Kubernetes API, and never rotated to tokens that are ephemeral,
+short lived, and automatically rotated.  By hardening these tokens and stabilizing
+their schema to match the semantics of OIDC ID tokens, we have enabled external consumers
+to validate these tokens for use cases beyond Kubernetes.  For example, it is possible to
+use a KSA token that is bound to a specific pod to fetch secrets from an external cloud
+provider vault without the need for any long lived secrets in the cluster.
+
+This KEP aims to make it possible for a similar 'secret-less' flow to be created for
+image pulls, which happen before the pod is running.  Today, admins are limited to image pull
+secrets that are stored directly in the Kubernetes API and thus are long lived and hard to rotate,
+or secrets that are managed at the Kubelet level via a Kubelet credential provider (meaning
+that any pod running on that node can access those images).  A pod should instead be able
+to use its own identity to pull images, i.e. we should enable image pull authorization to be
+tied to a particular workload while avoiding the need for long lived, persisted secrets.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+By moving to image pulls that are based on KSA tokens, we seek to:
+
+- Reduce secret management overhead for developers and admins
+- Reduce security risks associated with long lived secrets
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Allow workloads to pull images based on their own runtime identity without long lived / persisted secrets
+- Avoid needing a kubelet/node based identity to pull images
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We will expand the on-disk kubelet credential provider configuration to allow an
+optional `tokenAttribute` field to be configured.  When this field is not set, no KSA
+token will be sent to the plugin.  When it is set, the Kubelet will provision
+a token with the given audience bound to the current pod and its service
```

Comment:
will the token be time-bound? If so - how long?

Reply:
will it work nicely with things like stargz snapshotter where secret can be stored for a long time like: https://github.com/containerd/stargz-snapshotter/blob/main/docs/overview.md#cri-based-authentication? 

cc: @samuelkarp

Reply:
> will the token be time-bound? If so - how long?

Yes, the token will be time-bound. The duration will be fixed at 1h (?). If the cloud provider associates this token duration to the credential lifetime, we don't want it to be too short-lived. 1h seems reasonable to cover those scenarios to prevent frequent calls to cloud provider for token exchange.

Reply:
> will it work nicely with things like stargz snapshotter where secret can be stored for a long time like: https://github.com/containerd/stargz-snapshotter/blob/main/docs/overview.md#cri-based-authentication?

IIUC, the stargz snapshotter in `CRI-based authentication` gets the same registry credentials as the CRI from kubelet, so that wouldn't change here? The credentials could now be short-lived if the lifetime is tied to the KSA token lifetime but otherwise should work the same way. Please correct me if I'm wrong.

Reply:
stargz is not downloading the whole image in one go. So it may keep streaming pieces after `1h`. So it will end up with the stale credentials

Reply:
> stargz is not downloading the whole image in one go. So it may keep streaming pieces after 1h. So it will end up with the stale credentials

The credential provider plugin gets the PSAT and exchanges that for a registry credential. The registry credential is the one that `stargz` uses and it's not the PSAT right?

Reply:
correct. I think this can be compared to current behavior without specific PSAT: if a node wide credential expires while the image is being pulled with stargz, what does the runtime do? I'd expect the same here.

Reply:
oh, ok. So we are not giving time limited token to runtime. If this is the case, it will work

Reply:
I wonder if there is any test we can write that will validate that and ensure we are not breaking this scenario?

Reply:
The tests in kubelet credential provider wouldn't validate this because it fetches the registry credentials and passes it to CRI (no change in behavior). 

Tests in stargz would be good to confirm it works as expected when the registry credentials expire/it updates its cache when the credentials for the image change.
@SergeyKanzhelev do you know who can confirm if there are tests for this in `stargz`?